File: Python/Paginas/usuarios.py
import flet as ft
from Controladores_bases.ACbase_usuario import base_usuario

db = base_usuario()

def mostrar_usuarios(page):
    user_list = ft.ListView(expand=1, spacing=10, padding=20)
    username_input = ft.TextField(label="Nombre de usuario", border_color="#316938")
    password_input = ft.TextField(label="Contraseña", password=True, can_reveal_password=True, border_color="#316938")
    message = ft.Text(color="#316938")

    usuario_actual = None

    def cargar_usuarios():
        user_list.controls.clear()
        usuarios = db.listar_usuarios()
        for usuario in usuarios:
            user_list.controls.append(
                ft.Container(
                    ft.Row(
                        [
                            ft.Text(f"ID: {usuario[0]}", color="#316938"),
                            ft.Text(f"Nombre: {usuario[1]}", color="#316938"),
                            ft.IconButton(ft.icons.EDIT, icon_color="#316938", on_click=lambda e, usuario=usuario: editar_usuario(usuario)),
                            ft.IconButton(ft.icons.DELETE, icon_color="#316938", on_click=lambda e, usuario=usuario: borrar_usuario(usuario[1]))
                        ],
                        alignment=ft.MainAxisAlignment.SPACE_BETWEEN
                    ),
                    bgcolor="#F0F0F0",
                    border_radius=10,
                    padding=10
                )
            )
        page.update()

    def agregar_usuario(e):
        username = username_input.value
        password = password_input.value
        if db.registrar_usuario(username, password):
            message.value = "Usuario registrado exitosamente."
            cargar_usuarios()
        else:
            message.value = "Error: El nombre de usuario ya existe."
        username_input.value = ""
        password_input.value = ""
        page.update()

    def editar_usuario(user):
        nonlocal usuario_actual
        usuario_actual = user
        username_input.value = user[1]
        password_input.value = ""
        boton.text = "Actualizar usuario"
        boton.on_click = actualizar_usuario
        page.update()

    def actualizar_usuario(e):
        nonlocal usuario_actual
        new_password = password_input.value
        if usuario_actual and db.actualizar_usuario(usuario_actual[1], new_password):
            message.value = "Usuario actualizado exitosamente."
        else:
            message.value = "Error al actualizar el usuario."
        username_input.value = ""
        password_input.value = ""
        usuario_actual = None
        boton.text = "Agregar usuario"
        boton.on_click = agregar_usuario
        cargar_usuarios()
        page.update()

    def borrar_usuario(username):
        if db.eliminar_usuario(username):
            message.value = "Usuario eliminado exitosamente."
        else:
            message.value = "Error al eliminar el usuario."
        cargar_usuarios()
        page.update()


    boton = ft.ElevatedButton(
        "Agregar usuario",
        on_click = agregar_usuario,
        style=ft.ButtonStyle(color="white", bgcolor="#316938")
    )

    cargar_usuarios()

    return ft.View(
        "/users",
        [
            ft.AppBar(
                title=ft.Text("Administración de Usuarios", color="white"),
                bgcolor="#316938",
                center_title=True
            ),
            ft.Container(
                content=ft.Column([
                    user_list,
                    ft.Divider(color="#316938"),
                    ft.Text("Agregar/Editar Usuario:", size=20, color="#316938", weight=ft.FontWeight.BOLD),
                    username_input,
                    password_input,
                    ft.Row(
                        [boton],
                        alignment=ft.MainAxisAlignment.START,
                    ),
                    message,
                    ft.ElevatedButton(
                        "Regresar",
                        on_click=lambda _: page.go("/home"),
                        style=ft.ButtonStyle(color="white", bgcolor="#316938")
                    ),
                ], spacing=20),
                padding=20,
                bgcolor="#F0F0F0",
                expand=True
            )
        ],
        bgcolor="#F0F0F0"
    )

File: dart/Paginas/usuarios.dart
import 'package:flutter/material.dart';
import 'package:your_project_name/controllers/user_database_controller.dart';
import 'package:your_project_name/models/user.dart';

class UsersPage extends StatefulWidget {
  @override
  _UsersPageState createState() => _UsersPageState();
}

class _UsersPageState extends State<UsersPage> {
  final UserDatabaseController db = UserDatabaseController();
  final TextEditingController usernameController = TextEditingController();
  final TextEditingController passwordController = TextEditingController();
  String message = "";
  List<User> users = [];
  User? currentUser;

  @override
  void initState() {
    super.initState();
    loadUsers();
  }

  void loadUsers() async {
    List<Map<String, dynamic>> usersData = await db.listarUsuarios();
    setState(() {
      users = usersData.map((userData) => User.fromMap(userData)).toList();
    });
  }

  void addUser() async {
    String username = usernameController.text;
    String password = passwordController.text;
    User newUser = User(username: username, password: password);
    bool success = await db.registrarUsuario(newUser);
    setState(() {
      if (success) {
        message = "Usuario registrado exitosamente.";
        loadUsers();
      } else {
        message = "Error: El nombre de usuario ya existe.";
      }
      usernameController.clear();
      passwordController.clear();
    });
  }

  void editUser(User user) {
    setState(() {
      currentUser = user;
      usernameController.text = user.username;
      passwordController.clear();
    });
  }

  void updateUser() async {
    if (currentUser != null) {
      String newPassword = passwordController.text;
      bool success = await db.actualizarUsuario(currentUser!.username, newPassword);
      setState(() {
        if (success) {
          message = "Usuario actualizado exitosamente.";
          loadUsers();
        } else {
          message = "Error al actualizar el usuario.";
        }
        usernameController.clear();
        passwordController.clear();
        currentUser = null;
      });
    }
  }

  void deleteUser(String username) async {
    bool success = await db.eliminarUsuario(username);
    setState(() {
      if (success) {
        message = "Usuario eliminado exitosamente.";
        loadUsers();
      } else {
        message = "Error al eliminar el usuario.";
      }
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text("Administración de Usuarios"),
        backgroundColor: Color(0xFF316938),
      ),
      body: Container(
        color: Color(0xFFF0F0F0),
        child: Padding(
          padding: EdgeInsets.all(20),
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              Expanded(
                child: ListView.builder(
                  itemCount: users.length,
                  itemBuilder: (context, index) {
                    User user = users[index];
                    return Container(
                      margin: EdgeInsets.only(bottom: 10),
                      padding: EdgeInsets.all(10),
                      decoration: BoxDecoration(
                        color: Colors.white,
                        borderRadius: BorderRadius.circular(10),
                      ),
                      child: Row(
                        mainAxisAlignment: MainAxisAlignment.spaceBetween,
                        children: [
                          Text("ID: ${user.id}", style: TextStyle(color: Color(0xFF316938))),
                          Text("Nombre: ${user.username}", style: TextStyle(color: Color(0xFF316938))),
                          IconButton(
                            icon: Icon(Icons.edit, color: Color(0xFF316938)),
                            onPressed: () => editUser(user),
                          ),
                          IconButton(
                            icon: Icon(Icons.delete, color: Color(0xFF316938)),
                            onPressed: () => deleteUser(user.username),
                          ),
                        ],
                      ),
                    );
                  },
                ),
              ),
              Divider(color: Color(0xFF316938)),
              Text(
                "Agregar/Editar Usuario:",
                style: TextStyle(fontSize: 20, fontWeight: FontWeight.bold, color: Color(0xFF316938)),
              ),
              SizedBox(height: 10),
              TextField(
                controller: usernameController,
                decoration: InputDecoration(
                  labelText: "Nombre de usuario",
                  border: OutlineInputBorder(),
                  focusedBorder: OutlineInputBorder(
                    borderSide: BorderSide(color: Color(0xFF316938)),
                  ),
                ),
              ),
              SizedBox(height: 10),
              TextField(
                controller: passwordController,
                decoration: InputDecoration(
                  labelText: "Contraseña",
                  border: OutlineInputBorder(),
                  focusedBorder: OutlineInputBorder(
                    borderSide: BorderSide(color: Color(0xFF316938)),
                  ),
                ),
                obscureText: true,
              ),
              SizedBox(height: 10),
              ElevatedButton(
                child: Text(currentUser == null ? "Agregar usuario" : "Actualizar usuario"),
                onPressed: currentUser == null ? addUser : updateUser,
                style: ElevatedButton.styleFrom(
                  primary: Color(0xFF316938),
                  onPrimary: Colors.white,
                ),
              ),
              SizedBox(height: 10),
              Text(message, style: TextStyle(color: Color(0xFF316938))),
              SizedBox(height: 10),
              ElevatedButton(
                child: Text("Regresar"),
                onPressed: () => Navigator.of(context).pop(),
                style: ElevatedButton.styleFrom(
                  primary: Color(0xFF316938),
                  onPrimary: Colors.white,
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}
